### ISSUE TITLE
Consider eliminating potentially redundant views

### ISSUE THREAD

**Issue by luis-cortes** [kizitonwose/CalendarView#183]
Great work so far! I really appreciate you putting this together!

I was playing around with the layout inspector and sample app and was surprised to see how deeply nested the view hierarchy is and how many extra views are being generated. 

I believe the library is making too many `FrameLayouts` that are unnecessary in _some_ situations and at least one `LinearLayout` that, as far as I can tell, can be eliminated completely without any impact to users.

Removing the extra `LinearLayout`, if possible, would be a quick and easy win. 
Here's the block that creates it inside `CalendarAdapter.onCreateViewHolder`:
```
val monthBodyLayout = LinearLayout(context).apply {
  layoutParams = LinearLayout.LayoutParams(LP.WRAP_CONTENT, LP.WRAP_CONTENT)
  orientation = LinearLayout.VERTICAL
  id = bodyViewId
}
rootLayout.addView(monthBodyLayout)
```

If you remove the block above and make some other minor adjustments, the week views can be added directly to the `rootLayout`.

Removal of the extra `FrameLayout`s looks to be a little more tricky and I'd like to understand more about why it's there. However, if I understand what you're trying to do, I think there may be some better alternatives that accomplish what you're looking for.

**Comment by kizitonwose** [comment]
> If you remove the block above and make some other minor adjustments, the week views can be added directly to the rootLayout.

If this idea of yours could possibly morph into a pull request, that would be nice and very much appreciated.

---

> Removal of the extra FrameLayouts looks to be a little more tricky and I'd like to understand more about why it's there.

Kindly let me know which Framelayouts you're referring to and I can try to explain.

**Comment by luis-cortes** [comment]
> If this idea of yours could possibly morph into a pull request, that would be nice and very much appreciated.

I'd love to!

> Kindly let me know which Framelayouts you're referring to and I can try to explain.

Whoops! Can't believe I forgot to specify! Sorry about that. 

The `FrameLayouts` in question are in `DayHolder.inflateDayView()`
```
containerView = FrameLayout(parent.context).apply {
            // This will be placed in the WeekLayout(A LinearLayout) hence we
            // use LinearLayout.LayoutParams and set the weight appropriately.
            // The parent's wightSum is already set to 7 to accommodate seven week days.
            layoutParams = LinearLayout.LayoutParams(config.width, config.height, 1F)
            addView(dateView)
        }
```

**Comment by kizitonwose** [comment]
> The FrameLayouts in question are in DayHolder.inflateDayView()

This is used so we can set a custom ID to the main container without altering any ID the user may have set to the root layout provided via `dayViewRes`, see [`DayHolder` Line 54](https://github.com/kizitonwose/CalendarView/blob/8004f06e7dcdcd8a4babb74d6d2548b2c5769e2e/library/src/main/java/com/kizitonwose/calendarview/ui/DayHolder.kt#L54). Then we use this ID later in [CalendarLayoutManager Line 63](https://github.com/kizitonwose/CalendarView/blob/8004f06e7dcdcd8a4babb74d6d2548b2c5769e2e/library/src/main/java/com/kizitonwose/calendarview/ui/CalendarLayoutManager.kt#L63). I agree it can be removed, just not sure what the best approach would be. Maybe set the tag to the user-provided layout and use findViewWithTag later? I'm open to ideas.

**Comment by luis-cortes** [comment]
@kizitonwose So I just went ahead and removed the extra `FrameLayout` by doing this in `DayHolder`:
```
 fun inflateDayView(parent: LinearLayout): View {
        dateView = parent.inflate(config.dayViewRes).apply {
            // We ensure the layout params of the supplied child view is
            // MATCH_PARENT so it fills the parent container.
            layoutParams = LinearLayout.LayoutParams(config.width, config.height, 1F)
        }
        return dateView
    }

    fun bindDayView(currentDay: CalendarDay?) {
        this.day = currentDay
        if (!::viewContainer.isInitialized) {
            viewContainer = config.viewBinder.create(dateView)
        }

        if (currentDay != null) {
            if (!viewContainer.view.isVisible) {
                viewContainer.view.isVisible = true
            }
            config.viewBinder.bind(viewContainer, currentDay)
        } else if (!viewContainer.view.isGone) {
            viewContainer.view.isGone = true
        }
    }
```

As far as I can tell, doing this has no observable impact on any of the samples. 

I see what you were trying to do, but I'm not sure that users should have any expectation of using an ID in this manner. Are there any specific use cases where users would have to rely on an ID that couldn't be done differently? If not and you agree, I can put this together in a PR.

After getting rid of the extra `LinearLayout` and these `FrameLayout` is making the Layout Inspector view of this look really nice.

**Comment by kizitonwose** [comment]
> I see what you were trying to do, but I'm not sure that users should have any expectation of using an ID in this manner. Are there any specific use cases where users would have to rely on an ID that couldn't be done differently?

Considering we change the ID and the only solution would be for the user to introduce another layout, I would say it's still the same problem just masked differently. Plus I think that an internal layout means we can make further tweaks to it without enforcing any changes on the user's root layout. It's a Framelayout which is already optimized for this use case(having one child) so while it may not be very pretty on the layout inspector there wouldn't be any serious performance hit.

**Comment by luis-cortes** [comment]
I think I've done it 🙂 . I've managed to get rid of the `FrameLayout` without any impact to the user (as far as I can tell). All the calendars look the same and all the tests pass. PR incoming.

**Comment by kizitonwose** [comment]
Thought this would be closed since it's referenced in #186 but just remembered it's only for merges on the main branch. So closing this manually.
